plugins/node/opentelemetry-instrumentation-fastify/src/instrumentation.ts

```diff
@@ -0,0 +1,248 @@
+    const instrumentation = this;
+    return function onRequest(
+      request: FastifyRequest,
+      reply: FastifyReply,
+      done: HookHandlerDoneFunction
+    ) {
+      if (!instrumentation.isEnabled()) {
+        return done();
+      }
+      const rpcMetadata = getRPCMetadata(context.active());
+      const routeName = request.routerPath;
+      if (routeName && rpcMetadata?.type === RPCType.HTTP) {
+        rpcMetadata.span.updateName(`${request.method} ${routeName || '/'}`);
+      }
+      done();
+    };
+  }
+
+  private _wrapHandler(
+    pluginName: string,
+    hookName: string,
+    original: () => Promise<any> | void
+  ): () => Promise<any> | void {
+    const instrumentation = this;
+    return function (this: any, ...args): Promise<any> | void {
+      if (!instrumentation.isEnabled()) {
+        return original.apply(this, args);
+      }
+
+      const spanName = `${FastifyNames.MIDDLEWARE} - ${
+        original.name || ANONYMOUS_NAME
+      }`;
+
+      const span = instrumentation._startSpan(spanName, {
+        [AttributeNames.FASTIFY_TYPE]: FastifyTypes.MIDDLEWARE,
+        [AttributeNames.PLUGIN_NAME]: pluginName,
+        [AttributeNames.HOOK_NAME]: hookName,
+      });
+
+      return safeExecuteInTheMiddleMaybePromise(
+        () => {
+          return original.apply(this, args);
+        },
+        err => {
+          if (err) {
+            span.setStatus({
+              code: SpanStatusCode.ERROR,
+              message: err.message,
+            });
+            span.recordException(err);
+          }
+          span.end();
+        }
+      );
+    };
+  }
+
+  private _wrapAddHook(): (
+    original: FastifyInstance['addHook']
+  ) => () => FastifyInstance {
+    const instrumentation = this;
+    return function (
+      original: FastifyInstance['addHook']
+    ): () => FastifyInstance {
+      return function wrappedAddHook(this: any, ...args: any) {
+        if (!instrumentation.isEnabled()) {
+          return original.apply(this, args);
+        }
+
+        const name = args[0] as string;
+        const handler = args[1] as HandlerOriginal;
+        const pluginName = this.pluginName;
+        if (name === 'onRegister') {
+          return original.apply(this, [name as any, handler]);
+        }
+        return original.apply(this, [
+          name as any,
+          instrumentation._wrapHandler(pluginName, name, handler),
+        ]);
+      };
+    };
+  }
+
+  private _patchConstructor(
+    original: () => FastifyInstance
+  ): () => FastifyInstance {
+    const instrumentation = this;
+    return function (this: FastifyInstance, ...args) {
+      const app: FastifyInstance = original.apply(this, args);
+      app.addHook('onRequest', instrumentation._hookOnRequest());
+      app.addHook('preHandler', instrumentation._hookPreHandler());
+
+      instrumentation._wrap(app, 'addHook', instrumentation._wrapAddHook());
+
+      return app;
+    };
+  }
+
+  public _patchSend(span: Span) {
```

Comment:
@mcollina has mentioned to me that we should not wrap the .send() method, but rather use hooks instead since there is one after a response has been sent

Reply:
The reason is that execution is not complete when .send() is called. You can see the full lifecycle at https://www.fastify.io/docs/latest/Lifecycle/.

I would recommend using the `onResponse` hook.

Reply:
not really, I'm interested in time it took to process the whole inner function defined by the user. WIth hook there can be more operation. I'm catching precisely the beginning and end of this, the other hooks might just give the false information - hook might take longer because of some other operation, but here you will have precise information of user executed code.

Reply:
That's not a feasible risk for an end user - we do not offer any public way for checking if a hook is there.

It is a higher risk we change something in the internals of `reply.send()` that will break the monkeypatching.

Reply:
It doesn't matter - if the api changes then we update the instrumentation. The way it is patched now it also takes care of context propagation etc. etc. This works fine so I don't see any reason to change it tbh.

Reply:
> Monkeypatching adds overhead and decrease stability. You can achieve the same without slowing down the users by using `onSend`.

the number of operations will be exactly the same you still need to create the span, end it, set attributes etc. etc. so I don't see here any difference with regards to overhead. And if the api changes you will have to update the instrumentation anyway.

Reply:
I think @mcollina is saying that the external API is far less likely to change than the internal workings of the API, particularly in patch and minor releases.

I'm fine with whatever you decide to do, just resolve this conversation when you've decided.

Reply:
> It is a higher risk we change something in the internals of `reply.send()` that will break the monkeypatching.

reply is already typed the interface is called `FastifyReply`, if you change anything it means you will change also this interface. I would assume this will be a breaking change as well. Once we update to latest it will be also breaking then. And here are more problems with that especially the context propagation as well as enabling and disabling instrumentation and still being able to properly handle that, I don't see any reason to make it more complicated in that matter.

Reply:
As the author and lead developer of Fastify, I can only issue my best recommendation. Adding any monkeypatching when it is not needed adds unnecessary overhead and risks for the end users. It's some kind of last resort mechanism.

Anyhow, here is how DataDog implements it: https://github.com/DataDog/dd-trace-js/blob/master/packages/datadog-plugin-fastify/src/fastify.js.

Reply:
I'm glad you pointed DataDog here they are doing exactly the same - they patching the send method.
https://github.com/DataDog/dd-trace-js/blob/master/packages/datadog-plugin-fastify/src/fastify.js#L95
So why what DataDog is doing is fine but here is not fine anymore ?

Reply:
I missed that. Forget my comment - even if it's still correct.

Reply:
I think mcollina's suggestion will create a more robust implementation (in addition to being more efficient). On the other end, It will also make the instrumentation a bit more complex, as the span will need to be added as property on the req or reply to be available for the hook to call `span.end()`, instead of passing it as closure param which is simpler. It will also make the code flow a bit less obvious and harder to maintain as the `startSpan` and `span.end()` reside far away from each other (mentally and lexically).
I personally think neither of these two options is significantly preferred over the other and support whichever is finally chosen.